File: LeetCode/19.remove-nth-node-from-end-of-list.cpp
/*
 * @lc app=leetcode id=19 lang=cpp
 *
 * [19] Remove Nth Node From End of List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *removeNthFromEnd(ListNode *head, int n)
    {
        ListNode *fast = head, *slow = head, *tmp;
        for (int i = 0; i < n; i++)
        {
            fast = fast->next;
        }
        if (fast == nullptr)
        {
            tmp = head->next;
            delete head;
            return tmp;
        }
        while (fast->next)
        {
            slow = slow->next;
            fast = fast->next;
        }
        tmp = slow->next;
        slow->next = tmp->next;
        delete tmp;
        return head;
    }
};
// @lc code=end

File: LeetCode/19.remove-nth-node-from-end-of-list.py
#
# @lc app=leetcode id=19 lang=python3
#
# [19] Remove Nth Node From End of List
#


# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head, n):
        left = right = head

        for _ in range(n):
            right = right.next

        if right is None:
            return head.next

        while right.next:
            right = right.next
            left = left.next

        left.next = left.next.next

        return head


# @lc code=end
